test/base-fork/Morpho.t.sol

```diff
@@ -0,0 +1,272 @@
+
+    function test_morpho_deposit_notRelayer() external {
+        vm.expectRevert(abi.encodeWithSignature(
+            "AccessControlUnauthorizedAccount(address,bytes32)",
+            address(this),
+            RELAYER
+        ));
+        foreignController.depositERC4626(MORPHO_VAULT_USDS, 1_000_000e18);
+    }
+
+    function test_morpho_deposit_frozen() external {
+        vm.prank(freezer);
+        foreignController.freeze();
+
+        vm.prank(relayer);
+        vm.expectRevert("ForeignController/not-active");
+        foreignController.depositERC4626(MORPHO_VAULT_USDS, 1_000_000e18);
+    }
+
+    function test_morpho_deposit_rateLimited() external {
+        deal(Base.USDS, address(almProxy), 25_000_001e18);
+
+        assertEq(usdsVault.convertToAssets(usdsVault.balanceOf(address(almProxy))), 0);
+        assertEq(IERC20(Base.USDS).balanceOf(address(almProxy)),                    25_000_001e18);
+
+        vm.expectRevert("RateLimits/rate-limit-exceeded");
+        vm.prank(relayer);
+        foreignController.depositERC4626(MORPHO_VAULT_USDS, 25_000_001e18);
+
+        vm.prank(relayer);
+        foreignController.depositERC4626(MORPHO_VAULT_USDS, 25_000_000e18);
+
+        assertEq(usdsVault.convertToAssets(usdsVault.balanceOf(address(almProxy))), 25_000_000e18);
+        assertEq(IERC20(Base.USDS).balanceOf(address(almProxy)),                    1e18);
+    }
+
+    function test_morpho_withdraw_notRelayer() external {
+        vm.expectRevert(abi.encodeWithSignature(
+            "AccessControlUnauthorizedAccount(address,bytes32)",
+            address(this),
+            RELAYER
+        ));
+        foreignController.withdrawERC4626(MORPHO_VAULT_USDS, 1_000_000e18);
+    }
+
+    function test_morpho_withdraw_frozen() external {
+        vm.prank(freezer);
+        foreignController.freeze();
+
+        vm.prank(relayer);
+        vm.expectRevert("ForeignController/not-active");
+        foreignController.withdrawERC4626(MORPHO_VAULT_USDS, 1_000_000e18);
+    }
+
+    function test_morpho_redeem_notRelayer() external {
+        vm.expectRevert(abi.encodeWithSignature(
+            "AccessControlUnauthorizedAccount(address,bytes32)",
+            address(this),
+            RELAYER
+        ));
+        foreignController.redeemERC4626(MORPHO_VAULT_USDS, 1_000_000e18);
+    }
+
+    function test_morpho_redeem_frozen() external {
+        vm.prank(freezer);
+        foreignController.freeze();
+
+        vm.prank(relayer);
+        vm.expectRevert("ForeignController/not-active");
+        foreignController.redeemERC4626(MORPHO_VAULT_USDS, 1_000_000e18);
+    }
+
+}
+
+/**********************************************************************************************/
+/*** Success modes testing both USDS and USDC                                               ***/
+/**********************************************************************************************/
+
+contract MorphoUSDSSuccessTests is MorphoBaseTest {
```

Comment:
Break out by function, doing usds and usdc side by side

Reply:
MorphoDepositSuccessTests etc

Reply:
same here, i think there is a point of diminishing returns having 1 test per contract